### PR TITLE
Avoid using to-be-deprecated Data.List.NonEmpty.unzip

### PR DESCRIPTION
**Pull request by mixphix**
For CLC ticket [\#86](https://github.com/haskell/core-libraries-committee/issues/86)

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
[N/A] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
[N/A] The documentation has been updated, if necessary.
[N/A] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

Bonus points for added automated tests!

### CODE REVIEW

**Comment by ulysses4ever** [comment]
Can't we use `Data.Functor.unzip`? It doesn't have a since annotation, so I'm unsure if it fits into our support window. But that would be much preferable if it's possible imo.

**Comment by mixphix** [comment]
We'd have to restrict the bounds on `base` very heavily. I could add a comment about it in the source code, but I hesitate to touch more files than necessary.

**Comment by Kleidukos** [comment]
As said in the CLC ticket (https://github.com/haskell/core-libraries-committee/issues/86#issuecomment-1518662678) I'd be more comfortable if we had an automated way to prevent it from coming back, preferably with an HLint or Retrie rule.

**Comment by ulysses4ever** [comment]
@mixphix a comment nearby the code you added would be reasonable, I think.

**Review by ulysses4ever — APPROVED** [review]

File: Cabal/src/Distribution/Simple/BuildTarget.hs [hunk 1]
```diff
@@ -320,7 +321,9 @@ resolveBuildTarget pkg userTarget fexists =
   where
     classifyMatchErrors errs
       | Just expected' <- NE.nonEmpty expected
-                            = let (things, got:|_) = NE.unzip expected' in
+                            = let unzip' = fmap fst &&& fmap snd
+                                  -- ^ to work around deprecated NE.unzip
```
Comment (ulysses4ever) [inline Cabal/src/Distribution/Simple/BuildTarget.hs:325]: Well, by "comment" I mean an explanation that we can't use `Data.Functor.unzip` for compatibility reasons. The deprecation is eminent and sooner than we might perceive no one will remember of the generic `NE.unzip`. So, I'm not sure it's worth to put a comment about it.

Maybe, actually putting all the comments in the commit message is enough. Nothing will help too much with reading this code, and all the history belongs to commit messages imo.

Comment (mixphix) [reply Cabal/src/Distribution/Simple/BuildTarget.hs:325]: I'm happy to oblige. I'll force-push this branch with a single commit that explains why `Data.Functor.unzip` is unsuitable for the time being. Eventually `NE.unzip` may be reinstated, but the performance gain from that change will likely be minuscule.

Comment (ulysses4ever) [reply Cabal/src/Distribution/Simple/BuildTarget.hs]: Great, thanks!

**Comment by ocharles** [comment]
The use of `Data.List.NonEmpty.unzip` is entirely appropriate here, no? Is it just the warning that's problematic?

**Comment by mixphix** [comment]
The warning is causing the GHC pipeline to fail, see [here](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10189)

**Comment by mixphix** [comment]
The pipeline has at least `-Werror=deprecations`. Disabling warnings is a module-wide option that may hide future issues.

**Comment by ocharles** [comment]
Yea, you'd probably want to downgrade that selectively with CPP.

**Comment by mixphix** [comment]
You want me to use CPP to avoid a one-line local definition?

**Comment by ulysses4ever** [comment]
Hey @mixphix! Thanks a lot for the contribution! I see you requested my review but i already approved some time ago and that's about all i can do: the PR needs a second reviewer. We could tag some folks who are involved with Cabal but focused on other things and so didn't respond to this PR. E.g. @Kleidukos, @ffaf1, folks, I hope you don't mind me tagging you and perhaps you could find time to review this tiny improvement.

**Comment by Kleidukos** [comment]
With https://github.com/ndmitchell/hlint/pull/1500 I am now much more comfortable.

**Review by Kleidukos — APPROVED** [review]

**Review by ulysses4ever — APPROVED** [review]

**Comment by ulysses4ever** [comment]
@mixphix do you have permissions to add labels, and if so, are you ready to apply the merge-me label to hand it over to the merge bot?

**Comment by mixphix** [comment]
I don't have label permissions, but please go ahead!

**Comment by Kleidukos** [comment]
I gotchu

**Comment by Bodigrim** [comment]
What's up with Mergify here?

**Comment by Kleidukos** [comment]
Permissions stuff apparently, but let's merge this. :)

**Comment by ulysses4ever** [comment]
@mixphix the usual way with submodules, I think, is to cd into the submodule directory, do git pull, cd back, git add --all and commit...

**Pull request merged**